### ISSUE TITLE
ckeygen script should not prompt if 'pass' option is an empty string

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5998]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)| @teratorn reported|
|-|-|
|Trac ID|trac#5998|
|Type|defect|
|Created|2012-09-17 23:38:42Z|
|Branch|https://github.com/twisted/twisted/tree/ckeygen-no-passphrase-5998|

Driving ckeygen as a library is quite easy - though perhaps a little bit hackish. Nevertheless it is quite useful for having scripts that generate test keys, and the like. e.g.

```
from twisted.conch.scripts import ckeygen
opts = ckeygen.GeneralOptions()
opts['type'] = 'rsa'
opts['bits'] = 1024
opts['pass'] = '' # dont want a passphrases.
ckeygen.generateRSAkey(opts)
```

This doesn't work and still prompts for a passphrase due to the ckeygen script code checking the Truth value of 'pass' option instead of whether it is None (the default).

Attachments:

* [5998.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2af/2afe699e0928a1bfaf292d9fa1acf81b38438f9b/aaf26684af298e1637e4b184fe1f7316fe001450.patch) (2490 bytes) - added by magmatt on 2013-08-22 16:27:32Z - Add --no-passphrase option to ckeygen
* [5998-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2af/2afe699e0928a1bfaf292d9fa1acf81b38438f9b/bf2eb0ab3a46a1571e510e62572e91979f2b0531.patch) (2741 bytes) - added by magmatt on 2013-08-22 17:09:48Z - Patch including suggestions from exarkun and teratorn (in IRC)
* [5998-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2af/2afe699e0928a1bfaf292d9fa1acf81b38438f9b/d96bc22c27671d5c3e2ed0751c495d021f7ef150.patch) (2714 bytes) - added by magmatt on 2013-08-22 17:43:08Z - News change

<details><summary>Searchable metadata</summary>

```
trac-id__5998 5998
type__defect defect
reporter__teratorn teratorn
priority__normal normal
milestone__ 
branch__branches_ckeygen_no_passphrase_5998 branches/ckeygen-no-passphrase-5998
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ckeygen ckeygen
time__1347925122000000 1347925122000000
changetime__1377198921000000 1377198921000000
version__None None
owner__exarkun exarkun
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="magmatt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|magmatt set owner to magmatt|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="magmatt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|magmatt commented|
|-|-|

In discussing this with teratorn, it seems like a new option '--no-passphrase' is better than allowing an empty string in the case where a broken script incorrectly passes an empty string to --pass, thinking it passed a real passphrase.  (Though scripts don't actually think).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Thanks.

  1. Please write a news fragment explaining this user-facing change in behavior.
  1. `twisted.python.usage.Options` will supply `False` as the value for `no-passphrase`.  It seems the `.get(..., False)` is only required to make tests that don't pass a correct options instance in.  Perhaps the tests should be fixed?
  1. There are some comparisons of byte strings here.  It'd be great to spell the string literals as byte string literals.
  1. I think tests that make one assertion are better than tests that make several assertions.  If you can think of a way to reformulate this test so it only makes one assertion (perhaps by making it several tests), that'd be great.
  1. The test method docstring describes a scenario but doesn't explain the expected outcome.  Please add some more information to it.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="magmatt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|magmatt commented|
|-|-|

On my machine I have a patch that addresses 4 out of the 5 points.

Replying to [exarkun](#note_5):
>   2. `twisted.python.usage.Options` will supply `False` as the value for `no-passphrase`.  It seems the `.get(..., False)` is only required to make tests that don't pass a correct options instance in.  Perhaps the tests should be fixed?
Would it be acceptable to change it to `.get(...)`?  Many of the other functions in ckeygen.py use `.get(...)`.  Not that prior use means it's good -- just that it's consistent.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [39661]) Branching to 'ckeygen-no-passphrase-5998'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [39662]) Apply 5998-3.patch 

refs [#5998](https://github.com/twisted/twisted/issues/5998)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [39663]) Apply 5998-3.patch 

refs [#5998](https://github.com/twisted/twisted/issues/5998)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks!  I added the new option to the ckeygen man file as well (r39664).  Apart from that, this looks good.  I'll merge now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [39665]) Merge ckeygen-no-passphrase-5998

Author: magmatt
Reviewer: exarkun
Fixes: [#5998](https://github.com/twisted/twisted/issues/5998)

Add `--no-passphrase` option to Twisted Conch's `ckeygen` tool.  This allows
generation of a key with no passphrase and without a prompt for a passphrase.
